// bridge-sdk/lib/interfaces/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NETWORK_TYPE = exports.STATUS = exports.ACTION_TYPE = exports.TOKEN = exports.EXCHANGE_MODE = void 0;
var EXCHANGE_MODE;
(function (EXCHANGE_MODE) {
    EXCHANGE_MODE["ETH_TO_ONE"] = "eth_to_one";
    EXCHANGE_MODE["ONE_TO_ETH"] = "one_to_eth";
})(EXCHANGE_MODE = exports.EXCHANGE_MODE || (exports.EXCHANGE_MODE = {}));
var TOKEN;
(function (TOKEN) {
    TOKEN["BUSD"] = "busd";
    TOKEN["LINK"] = "link";
    TOKEN["ERC20"] = "erc20";
    TOKEN["ERC721"] = "erc721";
})(TOKEN = exports.TOKEN || (exports.TOKEN = {}));
var ACTION_TYPE;
(function (ACTION_TYPE) {
    ACTION_TYPE["depositOne"] = "depositOne";
    ACTION_TYPE["withdrawOne"] = "withdrawOne";
    ACTION_TYPE["getHRC20Address"] = "getHRC20Address";
    ACTION_TYPE["approveEthManger"] = "approveEthManger";
    ACTION_TYPE["lockToken"] = "lockToken";
    ACTION_TYPE["waitingBlockNumber"] = "waitingBlockNumber";
    ACTION_TYPE["mintToken"] = "mintToken";
    ACTION_TYPE["mintTokenRollback"] = "mintTokenRollback";
    ACTION_TYPE["approveHmyManger"] = "approveHmyManger";
    ACTION_TYPE["burnToken"] = "burnToken";
    ACTION_TYPE["waitingBlockNumberHarmony"] = "waitingBlockNumberHarmony";
    ACTION_TYPE["unlockToken"] = "unlockToken";
    ACTION_TYPE["unlockTokenRollback"] = "unlockTokenRollback";
    ACTION_TYPE["approveHRC20HmyManger"] = "approveHRC20HmyManger";
    ACTION_TYPE["approveHRC20EthManger"] = "approveHRC20EthManger";
    ACTION_TYPE["getERC20Address"] = "getERC20Address";
    ACTION_TYPE["lockHRC20Token"] = "lockHRC20Token";
    ACTION_TYPE["unlockHRC20Token"] = "unlockHRC20Token";
    ACTION_TYPE["burnHRC20Token"] = "burnHRC20Token";
    ACTION_TYPE["mintHRC20Token"] = "mintHRC20Token";
    ACTION_TYPE["unlockHRC20TokenRollback"] = "unlockHRC20TokenRollback";
    ACTION_TYPE["mintHRC20TokenRollback"] = "mintHRC20TokenRollback";
})(ACTION_TYPE = exports.ACTION_TYPE || (exports.ACTION_TYPE = {}));
var STATUS;
(function (STATUS) {
    STATUS["ERROR"] = "error";
    STATUS["SUCCESS"] = "success";
    STATUS["WAITING"] = "waiting";
    STATUS["IN_PROGRESS"] = "in_progress";
})(STATUS = exports.STATUS || (exports.STATUS = {}));
var NETWORK_TYPE;
(function (NETWORK_TYPE) {
    NETWORK_TYPE["ETHEREUM"] = "ETHEREUM";
    NETWORK_TYPE["BINANCE"] = "BINANCE";
})(NETWORK_TYPE = exports.NETWORK_TYPE || (exports.NETWORK_TYPE = {}));
