// bridge-sdk/lib/blockchain/out/HmyManagerERC721.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const abi = [
    {
        constant: false,
        inputs: [
            {
                internalType: 'address',
                name: 'oneToken',
                type: 'address',
            },
            {
                internalType: 'uint256[]',
                name: 'tokenIds',
                type: 'uint256[]',
            },
            {
                internalType: 'address',
                name: 'recipient',
                type: 'address',
            },
        ],
        name: 'burnTokens',
        outputs: [],
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        constant: true,
        inputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        name: 'mappings',
        outputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address',
            },
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function',
    },
];
exports.default = abi;
